types/hooks.d.ts

```diff
@@ -0,0 +1,328 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License.
+
+import { FunctionHandler } from '.';
+import { InvocationContext } from './InvocationContext';
+
+/**
+ * Represents a type which can release resources, such as event listening or a timer.
+ */
+export declare class Disposable {
+    /**
+     * Combine many disposable-likes into one. You can use this method when having objects with a dispose function which aren't instances of `Disposable`.
+     *
+     * @param disposableLikes Objects that have at least a `dispose`-function member. Note that asynchronous dispose-functions aren't awaited.
+     * @return Returns a new disposable which, upon dispose, will dispose all provided disposables.
+     */
+    static from(...disposableLikes: { dispose: () => any }[]): Disposable;
+
+    /**
+     * Creates a new disposable that calls the provided function on dispose.
+     * *Note* that an asynchronous function is not awaited.
+     *
+     * @param callOnDispose Function that disposes something.
+     */
+    constructor(callOnDispose: () => any);
+
+    /**
+     * Dispose this object.
+     */
+    dispose(): any;
+}
+
+/**
+ * A generic handler for hooks
+ */
+export type HookHandler = (context: HookContext) => void | Promise<void>;
+
+/**
+ * Base class for all hook context objects
+ */
+export declare class HookContext {
+    /**
+     * For testing purposes only. This will always be constructed for you when run in the context of the Azure Functions runtime
+     */
+    constructor(init?: HookContextInit);
+
+    /**
+     * Set a `propertyName` to some `value` to be shared with other hooks in the same scope (app-level vs invocation-level)
+     * @param propertyName The name of the property to set
+     * @param value The value to set
+     */
+    set(propertyName: string, value: unknown): void;
+
+    /**
+     * Get the value of a `propertyName` shared with other hooks in the same scope (app-level vs invocation-level)
+     * @param propertyName the name of the property to get
+     */
+    get(propertyName: string): unknown;
+
+    /**
+     * Set a `propertyName` to some `value` to be shared across scopes for all hooks
+     *
+     * @param propertyName The name of the property to set
+     * @param value The value to set
+     */
+    setGlobal(propertyName: string, value: unknown): void;
```

Comment:
So I don't want us to feel obligated to follow the core api, especially since we know we want to clean up the hook data. Taking a step back, how is `setGlobal` and `getGlobal` better than just using a global variable?

Aka rather than this:
```typescript
const onStartFunc: AppStartHandler = async (context: AppStartContext) => {
    context.setGlobal('globalKey', 'test123');
}
app.onStart(onStartFunc);

const onTerminateHook: AppTerminateHandler = async (context: AppTerminateContext) => {
    assert(context.getGlobal('globalKey') === 'test123');
};
app.onTerminate(onTerminateHook);
```

users could do this:
```typescript
let globalValue: string;

const onStartFunc: AppStartHandler = async (context: AppStartContext) => {
    globalValue = 'test123';
}
app.onStart(onStartFunc);

const onTerminateHook: AppTerminateHandler = async (context: AppTerminateContext) => {
    assert(globalValue === 'test123');
};
app.onTerminate(onTerminateHook);
```

If the user is defining their own global variables, they can more easily set the type (aka `string` instead of `unknown`), and they don't have to deal with the key (mispelling, conflicting with another key, etc.).

Lastly, I think it would simplify our lives if we only need to worry about invocation-scoped hook data, which IMO is something users can't accomplish on their own so it's much more valuable.